packages/ckeditor5-image/src/imagecaption/toggleimagecaptioncommand.js

```diff
@@ -0,0 +1,168 @@
+ *
+ * @extends module:core/command~Command
+ */
+export default class ToggleImageCaptionCommand extends Command {
+	/**
+	 * @inheritDoc
+	 */
+	refresh() {
+		const editor = this.editor;
+
+		// Only block images can get captions.
+		if ( !editor.plugins.has( ImageBlockEditing ) ) {
+			this.isEnabled = this.value = false;
+
+			return;
+		}
+
+		const selection = editor.model.document.selection;
+		const selectedElement = selection.getSelectedElement();
+
+		if ( !selectedElement ) {
+			const ancestorCaptionElement = getCaptionFromModelSelection( selection );
+
+			this.isEnabled = this.value = !!ancestorCaptionElement;
+
+			return;
+		}
+
+		// Only block images have captions.
+		this.isEnabled = isImage( selectedElement ) || isImageInline( selectedElement );
+
+		if ( !this.isEnabled ) {
+			this.value = false;
+		} else {
+			this.value = !!getCaptionFromImageModelElement( selectedElement );
+		}
+	}
+
+	/**
+	 * Executes the command.
+	 *
+	 *		editor.execute( 'toggleImageCaption' );
+	 *
+	 * @param {Object} [options] Options for the executed command.
+	 * @param {String} [options.focusCaptionOnShow] When true and the caption shows up, the selection will be moved into it straight away.
+	 * @fires execute
+	 */
+	execute( options = {} ) {
+		const { focusCaptionOnShow } = options;
+
+		this.editor.model.change( writer => {
+			if ( this.value ) {
+				this._hideImageCaption( writer );
+			} else {
+				this._showImageCaption( writer, focusCaptionOnShow );
+			}
+		} );
+	}
+
+	/**
+	 * Shows the caption of the `<image>` or `<imageInline>`. Also:
+	 *
+	 * * it converts `<imageInline>` to `<image>` to show the caption,
+	 * * it attempts to restore the caption content from the `caption` attribute,
+	 * * it moves the selection to the caption right away, it the `focusCaptionOnShow` option was set.
+	 *
+	 * @private
+	 * @param {module:engine/model/writer~Writer} writer
+	 */
+	_showImageCaption( writer, focusCaptionOnShow ) {
+		const model = this.editor.model;
+		const selection = model.document.selection;
+
+		let selectedImage = selection.getSelectedElement();
+		let newCaptionElement;
+
+		// Convert imageInline -> image first.
+		if ( isImageInline( selectedImage ) ) {
+			this.editor.execute( 'toggleImageType' );
+
+			// Executing the command created a new model element. Let's pick it again.
+			selectedImage = selection.getSelectedElement();
+		}
+
+		// Try restoring the caption from the attribute.
+		if ( selectedImage.hasAttribute( 'caption' ) ) {
+			newCaptionElement = Element.fromJSON( selectedImage.getAttribute( 'caption' ) );
+
+			// The model attribute is no longer needed if the caption was created out of it.
+			writer.removeAttribute( 'caption', selectedImage );
+		} else {
+			newCaptionElement = writer.createElement( 'caption' );
+		}
+
+		writer.append( newCaptionElement, selectedImage );
+
+		if ( focusCaptionOnShow ) {
+			writer.setSelection( newCaptionElement, 'end' );
```

Comment:
Maybe it would be more UX-friendly if we select the caption content so users could go straight to replacing the old caption that was hidden before.